src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -326,108 +312,69 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
-                    var receiver = new BoundImplicitReceiver(syntax, receiverType);
-                    BoundExpression boundAccess = BindIndexerOrIndexedPropertyAccess(syntax, receiver, indexerGroup, analyzedArguments, bindingDiagnostics);
-                    switch (boundAccess)
-                    {
-                        case BoundIndexerAccess boundIndexerAccess:
-                            if (boundIndexerAccess.ResultKind == LookupResultKind.Viable &&
-                                boundIndexerAccess.Indexer.GetMethod is { } getMethod &&
-                                IsAccessible(getMethod, ref useSiteInfo) &&
-                                TryLookupLengthOrCount(receiverType, lookupResult, out lengthProperty, ref useSiteInfo))
-                            {
-                                // PROTOTYPE(list-patterns) Can this be ever true? If so, move to if above
-                                Debug.Assert(!boundIndexerAccess.Indexer.IsStatic);
-                                ReportDiagnosticsIfObsolete(bindingDiagnostics, lengthProperty, syntax, hasBaseReceiver: false);
-                                GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, bindingDiagnostics, syntax: syntax);
-                                indexerAccess = BindIndexerDefaultArguments(boundIndexerAccess, BindValueKind.RValue, bindingDiagnostics);
-                                found = true;
-                                break;
-                            }
-                            found = false;
-                            break;
+            var success = BindLengthAndIndexerForListPattern(node, inputType, inputValEscape: ExternalScope, BindingDiagnosticBag.Discarded,
+                indexerAccess: out _, out var lengthAccess, receiverPlaceholder: out _, argumentPlaceholder: out _);
+            lengthProperty = success ? GetPropertySymbol(lengthAccess, out _, out _) : null;
+            return success;
+        }
 
-                        case BoundIndexOrRangePatternIndexerAccess boundIndexOrRangePatternIndexerAccess:
-                            lengthProperty = boundIndexOrRangePatternIndexerAccess.LengthOrCountProperty;
-                            patternSymbol = boundIndexOrRangePatternIndexerAccess.PatternSymbol;
-                            found = true;
-                            break;
+        private bool BindLengthAndIndexerForListPattern(SyntaxNode node, TypeSymbol inputType, uint inputValEscape, BindingDiagnosticBag diagnostics,
+            out BoundExpression indexerAccess, out BoundExpression lengthAccess, out BoundListPatternReceiverPlaceholder? receiverPlaceholder, out BoundListPatternIndexPlaceholder argumentPlaceholder)
+        {
+            if (inputType.IsDynamic())
+            {
+                Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
```

Comment:
Should we set hasErrors here?

Reply:
Sure
To be honest, I'm not completely sure about the importance of keeping diagnostics and hasError flags in bound tree in sync.

Reply:
It mainly is just used for suppressing later, potentially spurious warnings, so I think it would be good to set.